ElectricityBillApi.Tests/ElectricProviderTests.cs

```diff
@@ -0,0 +1,94 @@
+﻿using ElectricityBillApi.Models;
+using System;
+using Xunit;
+
+namespace ElectricityBillApi.Tests
+{
+    public class ElectricProviderTests
+    {
+        private ElectricProvider _sut;
+
+        public ElectricProviderTests()
+        {
+            _sut = new ElectricProvider();
+        }
+
+        [Fact]
+        public void CalculatePrice_CalculatesCorrectPrice()
+        {
+            // arrange
+            var loc1 = new Location()
+            {
+                X = 0,
+                Y = 0,
+                Z = 0
+            };
+            var loc2 = new Location()
+            {
+                X = 1,
+                Y = 1,
+                Z = 1
+            };
+            decimal plantPrice = 5;
+            var plant = new PowerPlant()
+            {
+                Location = loc1,
+                ElectricityPrice = plantPrice
+            };
+
+            // act
+            _sut.Subscribe(plant);
+            var actual = _sut.CalculatePrice(new Address() { Location = loc2 });
+            var expected = 3 * plantPrice;
+
+            // assert
+            Assert.Equal(expected, actual);
+        }
+
+        [Fact]
+        public void CalculatePrice_ThrowsIfNoPlantIsSubscribed()
+        {
+            // arrange
+
+            var loc2 = new Location()
+            {
+                X = 1,
+                Y = 1,
+                Z = 1
+            };
+
+            // act
+
+            // assert
+            Assert.ThrowsAny<ArgumentNullException>(() => _sut.CalculatePrice(new Address() { Location = loc2 }));
```

Comment:
This is fine, but in the last lesson we learned about delegates. You can use instead:

```cs
// Act
Action calculatePriceNoPlantSubscribed = () => _sut.CalculatePrice(new Address() { Location = loc2 })

// Assert
Assert.ThrowsAny<ArgumentNullException>(calculatePriceNoPlantSubscribed );
```

Reply:
Though not entirely correct (the act happens to be called in an assert) this still segregates the assertion and the act nicely.

Reply:
Never really thought about using delegates here :D tyvm, will definitely use them more here!